src/smt/smt_engine.h

```diff
@@ -113,1089 +123,1108 @@ namespace theory {
-   * corresponding sygus problem for generating such a solution.
+   * This method assumes d_subsolver has been initialized to do abduction
+   * problems.
    */
-  bool getAbduct(const Expr& conj, const Type& grammarType, Expr& abd);
-  /** Same as above, but without user-provided grammar restrictions */
-  bool getAbduct(const Expr& conj, Expr& abd);
+  bool getAbductInternal(Expr& abd);
 
   /**
-   * Get list of quantified formulas that were instantiated
+   * Helper method to obtain both the heap and nil from the solver. Returns a
+   * std::pair where the first element is the heap expression and the second
+   * element is the nil expression.
    */
-  void getInstantiatedQuantifiedFormulas( std::vector< Expr >& qs );
-   
+  std::pair<Expr, Expr> getSepHeapAndNilExpr();
+
   /**
-   * Get instantiations
+   * Get expanded assertions.
+   *
+   * Return the set of assertions, after expanding definitions.
    */
-  void getInstantiations( Expr q, std::vector< Expr >& insts );
-  void getInstantiationTermVectors( Expr q, std::vector< std::vector< Expr > >& tvecs );
+  std::vector<Expr> getExpandedAssertions();
 
-  /**
-   * Get an unsatisfiable core (only if immediately preceded by an
-   * UNSAT or VALID query).  Only permitted if CVC4 was built with
-   * unsat-core support and produce-unsat-cores is on.
+  /* Members -------------------------------------------------------------- */
+
+  /** Expr manager context */
+  context::Context* d_context;
+  /** User level context */
+  context::UserContext* d_userContext;
+  /** The context levels of user pushes */
+  std::vector<int> d_userLevels;
+
+  /** Our expression manager */
+  ExprManager* d_exprManager;
+  /** Our internal expression/node manager */
+  NodeManager* d_nodeManager;
+  /** The decision engine */
+
+  DecisionEngine* d_decisionEngine;
+  /** The theory engine */
+  TheoryEngine* d_theoryEngine;
+  /** The propositional engine */
+  prop::PropEngine* d_propEngine;
+
+  /** The proof manager */
+  ProofManager* d_proofManager;
+
+  /** An index of our defined functions */
+  DefinedFunctionMap* d_definedFunctions;
+
+  /** The SMT engine subsolver
+   *
+   * This is a separate copy of the SMT engine which is used for making
+   * calls that cannot be answered by this copy of the SMT engine. An example
+   * of invoking this subsolver is the get-abduct command, where we wish to
+   * solve a sygus conjecture based on the current assertions. In particular,
+   * consider the input:
+   *   (assert A)
+   *   (get-abduct B)
+   * In the copy of the SMT engine where these commands are issued, we maintain
+   * A in the assertion stack. To solve the abduction problem, instead of
+   * modifying the assertion stack to remove A and add the sygus conjecture
+   * (exists I. ...), we invoke a fresh copy of the SMT engine and leave the
+   * assertion stack unchaged. This copy of the SMT engine can be further
+   * queried for information regar
```

Comment:
You'll get a merge conflict with these (they were removed in https://github.com/CVC4/CVC4/pull/3182).

Reply:
Good catch, in this case it actually didn't produce a conflict and I might have missed it.